File: 03-march/17-count-matrices-with-top-left-element/solution.java
class Solution {
    public int countSubmatrices(int[][] grid, int k) {
        int m = grid.length;
        int n = grid[0].length;
        long[] col = new long[n];
        int ans = 0;

        for (int i = 0; i < m; i++) {
            long prefixSum = 0;
            for (int j = 0; j < n; j++) {
                col[j] += grid[i][j];
                prefixSum += col[j];   // sum of submatrix (0,0) to (i,j)
                if (prefixSum <= k) {
                    ans++;
                }
            }
        }

        return ans;
    }
}

File: 03-march/17-count-matrices-with-top-left-element/solution.py
from typing import List

class Solution:
    def countSubmatrices(self, grid: List[List[int]], k: int) -> int:
        m, n = len(grid), len(grid[0])
        col = [0] * n
        ans = 0

        for i in range(m):
            prefix_sum = 0
            for j in range(n):
                col[j] += grid[i][j]
                prefix_sum += col[j]   # sum of submatrix (0,0) to (i,j)
                if prefix_sum <= k:
                    ans += 1

        return ans
      